feat: Add custom http status code and headers support for cloud funct…

hajjarjoseph opened the pull request:
## Pull Request

  - Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
  - Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
  - Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

  ## Issue

  Closes: #9479

  ## Approach

  Modified `FunctionsRouter.createResponseObject` to detect `__httpResponse: true` in cloud function return values and pass through `status` and `headers` to the existing PromiseRouter infrastructure. Fully backwards compatible.
  
  edited new approach:
  
Added CloudResponse class and pass it as second argument to cloud functions (req, res). When res.status() or res.set() are called, FunctionsRouter.createResponseObject passes status and headers to the existing PromiseRouter infrastructure. The res argument is optional - existing functions using only (req) continue to work unchanged.

  ## Tasks

  - [x] Add tests
  - [x] Add changes to documentation (guides, repository pages, code comments)
  - [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
  - [ ] Add new Parse Error codes to Parse JS SDK

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Cloud Functions can optionally customize HTTP responses (set status codes and headers via an Express-like response object).

* **Documentation**
  * README updated with usage examples, chaining behavior, and security considerations for response headers; note that the response argument is optional.

* **Tests**
  * Extensive new tests covering status/header setting, validation, combined responses, error paths, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hajjarjoseph committed:
feat: add custom http status code and headers support for cloud functions

parseplatformorg commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/acinader/pr-checks/63c4e5fd-d589-4dfe-a28c-10cdb116daa5?refs=71fba5fd440c5195016fd3e24332194dc3b76f4e&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

mtrezza commented:
I'm not a fan of this implementation; it looks like manipulation a server-internal property. Exposing a response object alongside the request object would be a cleaner approach.

hajjarjoseph committed:
feat: add response argument to cloud functions for HTTP status and headers

hajjarjoseph commented:
> I'm not a fan of this implementation; it looks like manipulation a server-internal property. Exposing a response object alongside the request object would be a cleaner approach.



> I'm not a fan of this implementation; it looks like manipulation a server-internal property. Exposing a response object alongside the request object would be a cleaner approach.

Changed my approach:

Added CloudResponse class and pass it as second argument to cloud functions (req, res). When res.status() or res.set() are called, FunctionsRouter.createResponseObject passes status and headers to the existing PromiseRouter infrastructure. The res argument is optional - existing functions using only (req) continue to work unchanged.

  Example:
  ```
Parse.Cloud.define('myFunction', (req, res) => {
    res.status(201).set('X-Custom-Header', 'value');
    return { message: 'Created' };
  });
```

I do not want the bounty for this feature implementation, would just be happy if I get a tiny contribution to a project I have been using regularly!

hajjarjoseph committed:
feat: add response argument to cloud functions for HTTP status and headers

hajjarjoseph commented on src/Routers/FunctionsRouter.js:
```diff
@@ -9,6 +9,41 @@ import { jobStatusHandler } from '../StatusHandler';
 import _ from 'lodash';
 import { logger } from '../logger';
 
+class CloudResponse {
+  constructor() {
+    this._status = null;
+    this._headers = {};
+  }
+
+  status(code) {
+    if (typeof code !== 'number') {
+      throw new Error('Status code must be a number');
+    }
+    this._status = code;
+    return this;
+  }
+
+  set(name, value) {
+    if (typeof name !== 'string') {
+      throw new Error('Header name must be a string');
+    }
+    this._headers[name] = value;
+    return this;
+  }
+
+  hasCustomResponse() {
+    return this._status !== null || Object.keys(this._headers).length > 0;
+  }
+
+  getStatus() {
+    return this._status;
+  }
+
+  getHeaders() {
+    return this._headers;
+  }
+}
```
- Added validation: status code range (100-599), header name must be string, header value must be defined
  - Added defensive copy of headers object

Documentation:
  - Added "Cloud Functions HTTP Response" section to README with usage examples, API reference, and security considerations

Tests:
  - Added edge case tests for: multiple status() calls, multiple set() calls, invalid status codes, invalid header names/values

hajjarjoseph committed:
feat: add response argument to cloud functions for HTTP status and headers

hajjarjoseph commented on README.md:
````diff
@@ -771,6 +771,55 @@ Logs are also viewable in Parse Dashboard.
+  res.status(201);
+  return { id: 'abc123', message: 'Created' };
+});
+
+// Set custom headers
+Parse.Cloud.define('apiEndpoint', (req, res) => {
+  res.set('X-Request-Id', 'req-123');
+  res.set('Cache-Control', 'no-cache');
+  return { success: true };
+});
+
+// Chain methods
+Parse.Cloud.define('authenticate', (req, res) => {
+  if (!isValid(req.params.token)) {
+    res.status(401).set('WWW-Authenticate', 'Bearer');
+    return { error: 'Unauthorized' };
+  }
+  return { user: 'john' };
+});
+```
+
+### Response Methods
+
+| Method | Description |
+|--------|-------------|
+| `res.status(code)` | Set HTTP status code (e.g., 201, 400, 404). Returns `res` for chaining. |
+| `res.set(name, value)` | Set HTTP header. Returns `res` for chaining. |
+
+### Backwards Compatibility
+
+The `res` argument is optional. Existing cloud functions using only `(req) => {}` continue to work unchanged.
+
+### Security Considerations
+
+The `set()` method allows setting arbitrary HTTP headers. Be cautious when setting security-sensitive headers such as:
+- CORS headers (`Access-Control-Allow-Origin`, `Access-Control-Allow-Credentials`)
+- `Set-Cookie`
+- `Location` (redirects)
+- Authentication headers (`WWW-Authenticate`)
+
````
Done

hajjarjoseph commented on src/Routers/FunctionsRouter.js:
```diff
@@ -9,6 +9,47 @@ import { jobStatusHandler } from '../StatusHandler';
+class CloudResponse {
+  constructor() {
+    this._status = null;
+    this._headers = {};
+  }
+
+  status(code) {
+    if (typeof code !== 'number') {
+      throw new Error('Status code must be a number');
+    }
+    if (code < 100 || code > 599) {
+      throw new Error('Status code must be between 100 and 599');
+    }
+    this._status = code;
+    return this;
+  }
+
+  set(name, value) {
+    if (typeof name !== 'string') {
+      throw new Error('Header name must be a string');
+    }
+    if (value === undefined || value === null) {
+      throw new Error('Header value must be defined');
+    }
+    this._headers[name] = value;
+    return this;
+  }
+
+  hasCustomResponse() {
+    return this._status !== null || Object.keys(this._headers).length > 0;
+  }
+
+  getStatus() {
+    return this._status;
+  }
+
+  getHeaders() {
+    return this._headers;
+  }
+}
```
Done

mtrezza commented:
See https://github.com/parse-community/parse-server/pull/9980

hajjarjoseph commented:
> ## [Codecov](https://app.codecov.io/gh/parse-community/parse-server/pull/9979?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community) Report
> 
> ✅ All modified and coverable lines are covered by tests. ✅ Project coverage is 92.55%. Comparing base ([`d5e76b0`](https://app.codecov.io/gh/parse-community/parse-server/commit/d5e76b01db2b4eeb22a0bb5a04347a89209aa822?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community)) to head ([`71fba5f`](https://app.codecov.io/gh/parse-community/parse-server/commit/71fba5fd440c5195016fd3e24332194dc3b76f4e?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community)). ⚠️ Report is 1 commits behind head on alpha.
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/parse-community/parse-server/pull/9979?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community).
> 🚀 New features to boost your workflow:

@mtrezza did you check the changes here? Will you go with #9980? Thank you :-)

mtrezza commented:
@hajjarjoseph I think we'll go with https://github.com/parse-community/parse-server/pull/9980, it also contains a new express-style syntax for composing the response. Would you want to test out that PR before we merge - if you have a use case for it?

hajjarjoseph commented:
> @hajjarjoseph I think we'll go with #9980, it also contains a new express-style syntax for composing the response. Would you want to test out that PR before we merge - if you have a use case for it?
 
All good 👍
